src/tagmanager/tm_source_file.c

```diff
@@ -609,6 +609,24 @@ static gboolean tm_source_file_init(TMSourceFile *source_file, const char *file_
 	else
 		source_file->lang = tm_ctags_get_named_lang(name);
 
+	source_file->is_c_source = FALSE;
+
+	if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
+	{
+		const gchar **ext;
+		const gchar *common_src_exts[] =
+			{".c", ".C", ".cc", ".cp", ".cpp", ".cxx", ".c++", ".CPP", ".CXX", NULL};
+
+		for (ext = common_src_exts; *ext; ext++)
+		{
+			if (g_str_has_suffix(source_file->short_name, *ext))
+			{
+				source_file->is_c_source = TRUE;
+				break;
+			}
+		}
+	}
+
```

Comment:
wouldn't it be simpler to have `source_file->is_source`, initialize it to `TRUE` and use it for all languages, and make it `FALSE` for C and C++ if the extension doesn't match any of those?

```C
	source_file->is_source = TRUE;

	if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
	{
		const gchar *ext = strrchr(source_file->short_name, '.');

		if (! ext)
			source_file->is_source = FALSE;
		else
		{
			int i;
			const gchar *common_src_exts[] =
				{"c", "C", "cc", "cp", "cpp", "cxx", "c++", "CPP", "CXX"};

			for (i = 0; i < G_N_ELEMENTS(common_src_exts); i++)
			{
				if (strcmp(ext + 1, common_src_exts[i]) == 0)
					break
			}

			source_file->is_source = (i < G_N_ELEMENTS(common_src_exts));
		}
	}
```

Reply:
```
source_file->is_source = TRUE;
```

Yes, we could do that - makes probably more sense as the `is_source` flag is valid for all languages and not specific to C/C++.

```C
	if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
	{
		const gchar *ext = strrchr(source_file->short_name, '.');

		if (! ext)
			source_file->is_source = FALSE;
		else
		{
			int i;
			const gchar *common_src_exts[] =
				{"c", "C", "cc", "cp", "cpp", "cxx", "c++", "CPP", "CXX"};

			for (i = 0; i < G_N_ELEMENTS(common_src_exts); i++)
			{
				if (strcmp(ext + 1, common_src_exts[i]) == 0)
					break
			}

			source_file->is_source = (i < G_N_ELEMENTS(common_src_exts));
		}
	}
```
Does this code offer any benefit compared to what I wrote? I personally find my code a bit easier to read as you avoid the extra branch where `.` is not found and one also has to think a bit when looking at
```
source_file->is_source = (i < G_N_ELEMENTS(common_src_exts));
```
what exactly it means.

Reply:
> Yes, we could do that - makes probably more sense as the `is_source` flag is valid for all languages and not specific to C/C++.

Yeah, the main reason I suggest this would be to make the `init_tag()` check simply `tag->local = tag_entry->isFileScope && file->is_source` rather than having language-specific tests there as well.

> Does this code offer any benefit compared to what I wrote? I personally find my code a bit easier to read as you avoid the extra branch where `.` is not found

Merely premature optimization (comparison only compares useful things, and traverse the basename only once), so no, it doesn't really matter.

If it's only a readability issue, you could easily rewrite this ever so slightly like so:
```C
	source_file->is_source = TRUE;

	if (source_file->lang == TM_PARSER_C || source_file->lang == TM_PARSER_CPP)
	{
		const gchar *ext = strrchr(source_file->short_name, '.');

		source_file->is_source = FALSE;
		if (ext)
		{
			const gchar *common_src_exts[] =
				{"c", "C", "cc", "cp", "cpp", "cxx", "c++", "CPP", "CXX"};

			for (int i = 0; i < G_N_ELEMENTS(common_src_exts); i++)
			{
				if (strcmp(ext + 1, common_src_exts[i]) == 0)
				{
					source_file->is_source = TRUE;
					break;
				}
			}
		}
	}
```

anyway, either way is fine.